File: boot/admin/src/test/java/com/aechak/admin/support/FakeFileStorage.java
package com.aechak.admin.support;

import com.aechak.application.file.port.FileKey;
import com.aechak.application.file.port.FileStorage;
import com.aechak.application.file.port.IssueFileUrl;
import com.aechak.application.file.port.enums.FileType;
import com.aechak.application.file.port.enums.UploadPurpose;
import com.aechak.domain.support.Ulid;

/** 실 S3 대신 키 규칙만 흉내냄 — 어드민 테스트는 다운로드 URL 발급이 주 관심사. */
public class FakeFileStorage implements FileStorage {

    @Override
    public IssueFileUrl issueUploadUrl(UploadPurpose purpose, FileType fileType, long userId) {
        String key = FileKey.INSTANCE.tmpPrefixOf(userId, purpose) + Ulid.INSTANCE.generate() + "." + fileType.getExtension();
        return new IssueFileUrl("https://fake-presigned.local/" + key, key);
    }

    @Override
    public String promote(String tmpKey, UploadPurpose purpose) {
        return purpose.getPrefix() + "/" + tmpKey.substring(tmpKey.lastIndexOf('/') + 1);
    }

    @Override
    public String publicUrlOf(String key) {
        return "https://fake-cdn.local/" + key;
    }

    @Override
    public String issueDownloadUrl(String key, UploadPurpose purpose) {
        return "https://fake-download.local/" + key;
    }
}

File: boot/api/src/testFixtures/kotlin/com/aechak/api/support/FakeFileStorage.kt
package com.aechak.api.support

import com.aechak.application.file.port.FileKey
import com.aechak.application.file.port.FileStorage
import com.aechak.application.file.port.IssueFileUrl
import com.aechak.application.file.port.enums.FileType
import com.aechak.application.file.port.enums.UploadPurpose
import com.aechak.domain.support.Ulid

/**
 * S3를 호출하지 않고 운영 코드와 같은 키 형식으로 동작한다.
 * promotedKeys와 deletedKeys로 테스트 중 실행된 스토리지 작업을 확인할 수 있다.
 */
class FakeFileStorage : FileStorage {
    private val promoted = mutableListOf<String>()
    private val deleted = mutableListOf<String>()

    val promotedKeys: List<String> get() = promoted.toList()

    /** 테스트 중 삭제된 스토리지 키. */
    val deletedKeys: List<String> get() = deleted.toList()

    fun clearPromoted() = promoted.clear()

    override fun issueUploadUrl(
        purpose: UploadPurpose,
        fileType: FileType,
        userId: Long,
    ): IssueFileUrl {
        val key = "${FileKey.tmpPrefixOf(userId, purpose)}${Ulid.generate()}.${fileType.extension}"
        return IssueFileUrl("https://fake-presigned.local/$key", key)
    }

    override fun promote(
        tmpKey: String,
        purpose: UploadPurpose,
    ): String {
        val key = "${purpose.prefix}/${tmpKey.substringAfterLast('/')}"
        promoted += key
        return key
    }

    override fun publicUrlOf(key: String): String = "https://fake-cdn.local/$key"

    override fun delete(
        key: String,
        purpose: UploadPurpose,
    ) {
        deleted += key
    }
}
